src/releases.rs

```diff
@@ -33,71 +38,83 @@ pub fn assign_commits(log: &Logger, release_name: &str, previous_release: &str,
-
-        use schema::releases::dsl::*;
-        use models::Release;
-        use schema::commits::dsl::*;
-        use models::Commit;
-
-        let the_release = releases
-            .filter(version.eq(&release_name))
-            .filter(project_id.eq(release_project_id))
-            .first::<Release>(&connection)
-            .expect("could not find release");
-
-        // did we make this commit earlier? If so, update it. If not, create it
-        match commits.filter(sha.eq(&sha_name)).first::<Commit>(&connection) {
-            Ok(the_commit) => {
-                info!(log, "Commit {} already exists in the database, just assigning it to release {}", sha_name, the_release.version);
-                diesel::update(commits.find(the_commit.id))
-                    .set(release_id.eq(the_release.id))
-                    .get_result::<Commit>(&connection)
-                    .expect(&format!("Unable to update commit {}", the_commit.id));
-            },
-            Err(_) => {
-                let git_log = Command::new("git")
-                    .arg("-C")
-                    .arg(&path)
-                    .arg("--no-pager")
-                    .arg("show")
-                    .arg(r#"--format=%H %ae %an"#)
-                    .arg(&sha_name)
-                    .output()
-                    .expect("failed to execute process");
-
-                let git_log = git_log.stdout;
-                let git_log = String::from_utf8(git_log).unwrap();
-
-                let log_line = git_log.split('\n').nth(0).unwrap();
-
-                let mut split = log_line.splitn(3, ' ');
-
-                let the_sha = split.next().unwrap();
-                let the_author_email = split.next().unwrap();
-                let the_author_name = split.next().unwrap();
-
-                info!(log, "Creating commit {} for release {}", the_sha, the_release.version);
-
-                let author = ::authors::load_or_create(&connection, &the_author_name, &the_author_email);
-                ::commits::create(&connection, &the_sha, &author, &the_release);
-            },
-        };
-    }
+type Sha<'a> = &'a str;
+type Email<'a> = &'a str;
+type Name<'a> = &'a str;
+type AuthorId = i32;
+
+/// Finds or creates all authors from a git log, and returns the given shas
+/// zipped with the id of the author in the database.
+fn authors_by_sha<'a>(conn: &PgConnection, git_log: Vec<(Sha<'a>, Email, Name)>)
+    -> QueryResult<Vec<(Sha<'a>, AuthorId)>>
+{
+    let new_authors = git_log.iter().map(|&(_, email, name)| {
+        NewAuthor { email: email, name: name }
+    }).collect();
+    let author_ids = ::authors::find_or_create_all(conn, new_authors)?
+        .into_iter()
+        .map(|author| ((author.email, author.name), author.id))
+        .collect::<HashMap<_, _>>();
+    Ok(git_log.into_iter()
+        .map(|(sha, email, name)| (sha, author_ids[&(email.into(), name.into())]))
```

Comment:
I'm really mad that I couldn't index a `HashMap<(String, String)` with a value of `(&str, &str)`